s2/cellunion.go

```diff
@@ -130,15 +130,48 @@ func CellUnionFromIntersectionWithCellID(x CellUnion, id CellID) CellUnion {
 	return cu
 }
 
+// cellUnionDifferenceInternal adds (xid - y) to the CellUnion. It subdivides
+// xid when there is partial overlap, narrowing y before recursing.
+func (cu *CellUnion) cellUnionDifferenceInternal(xid CellID, y CellUnion) {
+	var lo, hi int
+	if len(y) > 0 {
+		idMin := xid.RangeMin()
+
+		// Find first cell that could overlap.
+		lo = sort.Search(len(y), func(i int) bool {
+			return y[i].RangeMax() >= idMin
+		})
+
+		idMax := xid.RangeMax()
+
+		// Find first cell past our range.
+		hi = lo + sort.Search(len(y)-lo, func(i int) bool {
+			return y[(lo+i)].RangeMin() > idMax
+		})
+	}
+
+	if lo >= hi {
+		*cu = append(*cu, xid)
+		return
+	}
+
+	y = y[lo:hi]
```

Comment:
By reducing the size of "y" we're narrowing the search in the call to "ContainsCellID" and in recursive calls.